### DISCUSSION TITLE
Exporting constants in an emitter?

### DISCUSSION THREAD

**Discussion by jceipek** [microsoft/typespec#4247] [Q&A]
I'm trying to make a TypeSpec emitter that emits constant values that come from TypeSpec definitions.

Naively, I expected `const`s to work for this purpose, but they seem very similar to `alias`es in that I can't seem to access them in an emitter.

For example, the `defaultValue` of `field` here is numeric value 10:
```typespec
const MY_CONSTANT: safeint = 10;

model Test {
  field: safeint = MY_CONSTANT;
}
```
Trying to decorate a `const` produces the TypeSpec error: "Cannot decorate const statement.TypeSpec(invalid-decorator-location)".

I also tried to use `enum`s for the same goal, but `enum` values can't be used in numeric contexts:

```typespec
enum Constants {
  MY_CONSTANT: 10,
}

model Test {
  b: safeint = Constants.MY_CONSTANT; // "Type 'Constants.MY_CONSTANT' is not assignable to type 'safeint'TypeSpec(unassignable)"
}
```

Is there another way to do this?

**Comment by timotheeguerin** [comment]
Yeah you cannot decorate `const` as those are things resolved internally in the typespec compiler.

I think the enum case however should maybe work, could be worth fillying an issue for it.

It however works with a union like that 

```tsp
union Constants {
  MY_CONSTANT: 10,
}

model Test {
  b: safeint = Constants.MY_CONSTANT;
}

```
[Playground](https://typespec.io/playground?c=dW5pb24gQ29uc3RhbnRzIHsKICBNWV9DT05TVEFOVDogMTAsCn0KCm1vZGVsIFRlc3TFI2I6IHNhZmVpbnQgPco%2BLss6Owp9Cg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%22options%22%3A%7B%22%40typespec%2Fjson-schema%22%3A%7B%22file-type%22%3A%22json%22%7D%7D%7D)

**Reply by jceipek** [reply]
Thanks, @timotheeguerin! I could write a decorator to convert that into an enum or bare constants.

Is there a way to use the value in a doc comment template as well?
This obviously doesn't work, but maybe something similar?

```
union Constants {
  MY_CONSTANT: 10,
}

model Test {
  @doc("If b is set to the default of {Constants.MY_CONSTANT}, then etc...", Constants)
  b: safeint = Constants.MY_CONSTANT;
}
```

I find the error `Argument of type 'Constants' is not assignable to parameter of type '{}'` confusing.

**Reply by jceipek** [reply]
I filed a bug for the original issue here: https://github.com/microsoft/typespec/issues/4302
